src/sync/app.cpp

```diff
@@ -611,5 +627,107 @@ void App::log_out(std::function<void (Optional<AppError>)> completion_block) con
     }, handler);
 }
 
+void App::do_authenticated_request(Request request,
+                                   std::function<void (Response)> completion_block) const
+{
+    auto handler = [completion_block, request, this](const Response& response) {
+        if (auto error = check_for_errors(response)) {
+            App::handle_auth_failure(error.value(), response, request, completion_block);
+        } else {
+            completion_block(response);
+        }
+    };
+    
+    request.headers = get_request_headers(true, request.uses_refresh_token);
+    m_config.transport_generator()->send_request_to_server(request, handler);
+}
+
+void App::handle_auth_failure(const AppError& error,
+                              const Response& response,
+                              const Request& request,
+                              std::function<void (Response)> completion_block) const
+{
+    
+    auto transport_generator = m_config.transport_generator();
+    auto access_token_handler = [&transport_generator,
+                                 completion_block,
+                                 request,
+                                 response](const Optional<AppError>& error) {
+        if (!error) {
+            transport_generator->send_request_to_server(request, completion_block);
+        } else {
+            // pass the error back up the chain
+            completion_block(response);
+        }
+    };
+    
+    // Only handle auth failures
+    if (error.error_code.value() != 401) {
+        completion_block(response);
+        return;
+    }
+        
+    App::refresh_access_token_if_needed(request, access_token_handler);
```

Comment:
Some things are missing from this logic.

In the past, we automatically started a refresh of the token 10-30 seconds before it expired, in that case, users should hopefully not experience any "hiccups" in the sync connection.

Also the `refresh_access_token_if_needed` doesn't seem to retry itself if it fails. Before all bindings had an exponential backoff algorithm for this, which continuously retried getting the token automatically.

Not sure if it is useful, but the current Java logic is here: https://github.com/realm/realm-java/blob/master/realm/realm-library/src/objectServer/java/io/realm/internal/network/ExponentialBackoffTask.java#L27 and getting the access token logic is here: https://github.com/realm/realm-java/blob/master/realm/realm-library/src/objectServer/java/io/realm/SyncSession.java#L776

Reply:
Agreed that `refresh_access_token_if_needed` should retry itself if it fails.

However, we're trying to explicitly avoid doing any kind of threading in the App Layer. So if we want to do pre-emptive refreshes, it may have to be something the SDKs hook into.

Re: hiccups in the sync connection, I believe @nirinchev said we don't have to refresh the connection anymore? Or did I misunderstand that conversation?

Reply:
I don't have any opinion on exactly how it is done. If we want to avoid doing threading in OS, then yes, the SDK's need to be involved somehow.

I strongly assume that access tokens still expire? With refresh, I meant getting a new one before the old one expired.

Reply:
Resolved offline. Refreshing a sync session pre-emptively is no longer needed (or at all): "we don’t need to refresh the session. if the client sends the command, the server will accept it and do nothing with it. so it’s safe to remove it but also doesn’t break anything by having it there in the meantime"